pio/ppm/ppm.pio

```diff
@@ -0,0 +1,79 @@
+;
+; Copyright (c) 2020 Raspberry Pi (Trading) Ltd.
+;
+; SPDX-License-Identifier: BSD-3-Clause
+;
+
+.program ppm
+.side_set 1
+
+    set pindirs, 1  side 0; Set pin to output
+.wrap_target
+    out x, 16 side 0
+delay0:
+    jmp x-- delay0 side 0
+    out x, 16 side 1
+delay1:
+    jmp x-- delay1 side 1
+    .wrap
+
+% c-sdk {
+#include "hardware/pio.h"
+#include "hardware/clocks.h"
+#include "ppm.pio.h"
+
+
+#define PPM_CHANNELS (8)
+struct {
+    volatile PIO pio;
+    // 16 MSB - strobe length in microsec
+    // 16 LSB - gap length in microsec
+    volatile uint32_t ch_values[PPM_CHANNELS + 1];
+    volatile uint value_idx;
+} ppm_data;
+
+void ppm_handler() {
+    while (!pio_sm_is_tx_fifo_full(ppm_data.pio, 0)) {
+        pio_sm_put(ppm_data.pio, 0, ppm_data.ch_values[ppm_data.value_idx]);
+        ppm_data.value_idx = (ppm_data.value_idx + 1) % (PPM_CHANNELS + 1);
+    }
+}
+
+static inline void ppm_program_init(PIO pio, uint pin) {
+    ppm_data.pio = pio == NULL ? pio0 : pio;
+    ppm_data.ch_values[0] = 499 * 0x10000 +  4999;
```

Comment:
_Perhaps_ it'd be useful to have a macro which did
```c
#define BUILD_CH_VALUE(pulse) (((500 - 1) << 16) + (pulse - 1))
```
and then you could replace
```c
ppm_data.ch_values[0] = 499 * 0x10000 +  4999;
```
(and other similar lines) with
```c
ppm_data.ch_values[0] = BUILD_CH_VALUE(5000);
```
?